### PR TITLE
subscriber: drafted interface

### PR DESCRIPTION
**Pull request by mxmCherry**
Will do the in-mem consumer (+ README consuming recipes) in next PR - want to agree on interface first.

### CODE REVIEW

**Comment by mxmCherry** [comment]
@dim could you, please, push smth like `feature/consumer` so I could PR against it to accumulate all the consumer changes in one branch?

File: subscribe.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+// Subscriber defines the main subscriber interface.
+type Subscriber interface {
+	// Subscribe subscribes for topic messages and blocks till context is cancelled or error occurs.
+	//
+	// Handler is provided an atomic message batch (entire batch either succeeds or fails).
+	// Batch is guaranteed to contain at least 1 message.
+	Subscribe(
+		ctx context.Context,
+		topic string,
+		handler func([]Message) error,
```
Comment (mxmCherry) [inline subscribe.go:33]: `[]Message` is not an overkill - I'd like to make atomic-batch a base approach for consuming. Probably, we can have simpler `Subscriber` wrapper with single-message handler.

I did not extract handler func type not to scroll the docs when using bps.Consumer for it.

File: subscribe.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+		opts *SubscribeOptions,
```
Comment (mxmCherry) [inline subscribe.go:34]: Happy to turn it into `opts ...func(*SubscribeOptions)` for neat(er) API, should I though?

File: subscribe.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+	"net/url"
+	"sync"
+)
+
+var (
+	subReg   = map[string]SubscriberFactory{}
+	subRegMu sync.Mutex
+)
+
+// SubscribeOptions defines subscribe options.
```
Comment (dim) [inline subscribe.go:15]: `// Please note that not every option may be supported by the underlying implementation`

File: subscribe.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+type SubscribeOptions struct {
+	// BatchSize holds desired message batch size.
```
Comment (dim) [inline subscribe.go:17]: `The number of messages fetched from the server in a single request`

File: subscribe.go [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+	// Default: 1.
```
Comment (dim) [inline subscribe.go:18]: I wouldn't, let's not default and say: `Default: implementation specific`.

File: subscribe.go [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+	BatchSize int
+}
+
+// Normalize normalises subscriber options by applying defaults.
+func (o *SubscribeOptions) Normalize() *SubscribeOptions {
```
Comment (dim) [inline subscribe.go:23]: remove for now

File: subscribe.go (hunk 1 above)
Comment (dim) [reply subscribe.go:44]: let's add a `Handler` interface with `ProcessMessage(contenxt.Context, Msg) error` and a `HandlerFunc`. We should also not operate on `Message` struct but on an `Msg` interface. One is for producing the other for consuming.

File: subscribe.go [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+		handler func([]Message) error,
+		opts *SubscribeOptions,
+	) error
+	// Close closes the subscriber connection.
+	Close() error
+}
+
+// NewSubscriber inits to a subscriber via URL.
+//
+//   sub, err := bps.NewSubscriber(context.TODO(), "kafka://10.0.0.1:9092,10.0.0.2:9092,10.0.0.3:9092/namespace")
```
Comment (dim) [inline subscribe.go:53]: can we use the same URL pattern for producers and consumers?

File: subscribe_test.go [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+	It("should panic when trying to register the same scheme 2x", func() {
+		Expect(func() {
+			bps.RegisterSubscriber("mem", nil)
+		}).To(Panic())
+	})
+
+	It("should allow to init subscribers by URL", func() {
+		sub, err := bps.NewSubscriber(ctx, "mem://test.host/path")
+		Expect(err).NotTo(HaveOccurred())
+		Expect(sub).To(BeAssignableToTypeOf(dummySubscriber{}))
```
Comment (dim) [inline subscribe_test.go:31]: `defer Close()` just for "demo purposes"

File: subscribe.go (hunk 7 above)
Comment (mxmCherry) [reply subscribe.go:53]: @dim exactly what I'm doing, check this: https://github.com/bsm/bps/pull/3

File: subscribe.go (hunk 1 above)
Comment (mxmCherry) [reply subscribe.go:44]: @dim maybe better this?

```go
type Message interface { ... }
type ProduceMessage struct { ... }
```

Msg/Message are kinda confusing / don't tell much (IMO).

File: subscribe.go (hunk 3 above)
Comment (mxmCherry) [reply subscribe.go:15]: @dim don't think we need it at all if we're going to switch to single-message consuming (so single-message ack-ing as well)

File: subscribe.go (hunk 1 above)
Comment (mxmCherry) [reply subscribe.go]: I'm doing this ^^

File: subscribe.go [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+
+// ----------------------------------------------------------------------------
+
+// Subscriber defines the main subscriber interface.
+type Subscriber interface {
+	// Subscribe subscribes for topic messages and blocks till context is cancelled or error occurs.
+	//
+	// Handler is provided an atomic message batch (entire batch either succeeds or fails).
+	// Batch is guaranteed to contain at least 1 message.
+	Subscribe(ctx context.Context, topic string, handler Handler) error
```
Comment (mxmCherry) [inline subscribe.go:44]: Question: probably, better have this? cc @dim 

```go
csm, _ := NewConsumer("kafka://.../namespace?topics=foo,bar,baz")
_ := csm.Subscribe(ctx, handler)
```

Comment (dim) [reply subscribe.go:44]: so I can see why `NewConsumer` should have `ctx` (as this is where we do network IO), not so sure about `Subscribe`. let's not overdo it with context

File: subscribe.go [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+	return f(ctx, msg)
+}
+
+// ----------------------------------------------------------------------------
+
+// Subscriber defines the main subscriber interface.
+type Subscriber interface {
+	// Subscribe subscribes for topic messages and blocks till context is cancelled or error occurs.
+	//
+	// Handler is provided an atomic message batch (entire batch either succeeds or fails).
```
Comment (dim) [inline subscribe.go:42]: fixme: comment

File: subscribe.go [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+// Handle handles a single message.
+func (f HandlerFunc) Handle(ctx context.Context, msg SubMessage) error {
+	return f(ctx, msg)
+}
+
+// ----------------------------------------------------------------------------
+
+// Subscriber defines the main subscriber interface.
+type Subscriber interface {
+	// Subscribe subscribes for topic messages and blocks till context is cancelled or error occurs.
```
Comment (dim) [inline subscribe.go:40]:
```
... blocks until an error occurs.
```

one more consideration: handlers need to be able to exit gracefully, should we add an `ErrStop` or should we let the user deal with it:

```go
var MyOwnErr = errors.New(..)

err := client.Subscribe("topic", bps.HanderFunc(func(msg SubMessage) error {
  return MyOwnErr
})
if err == MyOwnErr {
   return nil // ugly, I think, better add ErrStop?
}
```

Comment (mxmCherry) [reply subscribe.go:40]: I'll do `bps.Done` for it - no `Err` prefix at all, just `bps.Done` (to be really nicely readable).

File: subscribe.go (hunk 9 above)
Comment (mxmCherry) [reply subscribe.go:44]: What if we want to dynamically subscribe? Like, to keep subscription not for the whole process run time, but just for some time?

Having context as a "unsubscribe" option would be useful then.

Btw, @dim what about the original question - binding topics in constructor and not in Subscribe? Overthinking?

**Pull request merged**